### PR TITLE
Feat/2965 db seed impl

### LINKED ISSUE #2965: As tech lead, I want a database seed implemented for testing
**Description:**
Given our occasional setbacks resulting from database inconsistencies that result in deployment failures that back up our release schedules, we should mitigate this with more thorough testing.

**Acceptance Criteria:** 

- [x] We have a maintainable and extensible toolset
- [ ] Seeding a database is performant ( sub hour )
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
- [x] Create a tool/mechanism for generating "random" and internally consistent data
- [x] #3168
- [x] Tool will "fuzz" or generate out of range values to intentionally create issues [ stretch ]
- [ ] #3175
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**

**Supporting Documentation:**

**Open Questions:**

- Can we work with Alex to scrub PII from the prod postgres dataset?
- How can this be integrated into unit/integration tests? Pipelines?
- What drift exists between postgres migrations and ES documents? Is that prone to human error?
- Can we leverage co-pilot or other AI for generation?

### PR DESCRIPTION
**Pull request by andrew-jameson**
## Summary of Changes
Pull request closes #2965

## How to Test

```
task drop-db
task init-backend
task backend-exec-seed-db
```

1. Open http://localhost:3000/ and sign in.
1. Proceed to Django Admin Console and ensure hundreds of datafiles associated to various STTs exist.
2. Ensure a postegres dump file is in `TANF-app/tdrs-backend/tdrs_db_seed.pg`

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [ ] We have a maintainable and extensible toolset
- [ ] Seeding a database is performant ( sub hour )
- [x] Create a tool/mechanism for generating "random" and internally consistent dat
- [x] Tool will "fuzz" or generate out of range values to intentionally create issues [ stretch ]
- [X] Create django command such that tool can be pointed at deployed environments [ stretch ]
+ [ ] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [ ] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-backend/tdpservice/parsers/management/commands/seed_db.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+class Command(BaseCommand):
+    """Command class."""
+
+    help = "Populate datafiles, records, summaries, and errors for all STTs."
+
+    def handle(self, *args, **options):
+        """Populate datafiles, records, summaries, and errors for all STTs."""
+   
+        for stt in STT.objects.all():  # filter(id__in=range(1,2)):
```
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/parsers/management/commands/seed_db.py]: I think `all` might be excessive here, might take in an argument that utilizes the range in the comment to have a more limited subset of STTs.

File: tdrs-backend/tdpservice/parsers/management/commands/seed_db.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+                        line += stt.stt_code
+                    else:
+                        line += '000'
+                elif field.name == 'program_type':
+                    line += prog_type
+                elif field.name == 'edit':
+                    line += '1'
+                elif field.name == 'encryption':
+                    line += 'E'
+                elif field.name == 'update':
```
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/parsers/management/commands/seed_db.py]: I really wish I could condense this, but the value-add of this script isn't to generate a bunch of files that fail at the first line and get auto-rejected. Having a good header is paramount but there are enough differences in the headers of various files that I can no longer just use the hardcoded one-liner I had previously.

File: tdrs-backend/tdpservice/parsers/management/commands/seed_db.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+from tdpservice.parsers.test.factories import DataFileSummaryFactory
+from tdpservice.scheduling.parser_task import parse as parse_task
+from tdpservice.stts.models import STT
+from tdpservice.users.models import User
+from tdpservice.parsers.row_schema import RowSchema
+
+logger = logging.getLogger(__name__)
+
+# https://faker.readthedocs.io/en/stable/providers/baseprovider.html#faker.providers.BaseProvider
+""" class FieldFaker(faker.providers.BaseProvider):..."""
```
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/parsers/management/commands/seed_db.py]: Should probably delete this, but wanted to leave myself this breadcrumb.

File: tdrs-backend/tdpservice/parsers/management/commands/seed_db.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        section = text_dict['section']  # A
+        models_in_section = get_program_models(prog_type, section)
+        temp_file = ''
+
+        cal_year, cal_quarter = fiscal_to_calendar(sub_year, 'Q{}'.format(sub_quarter))
+        temp_file += make_HT(header, prog_type, section, cal_year, cal_quarter, stt)
+
+        # iterate over models and generate lines
+        for _, model in models_in_section.items():
+            if long_section in ['Active Case Data', 'Closed Case Data', 'Aggregate Data', 'Stratum Data']:
```
Comment (elipe17) [inline tdrs-backend/tdpservice/parsers/management/commands/seed_db.py]: Is SSP covered with this too?

Comment (andrew-jameson) [reply tdrs-backend/tdpservice/parsers/management/commands/seed_db.py]: No, it definitely wouldn't. I'll get that fixed up

**Review by elipe17 — CHANGES_REQUESTED** [review]
This PR led me to find a very small but impactful error in another area of our parser code that causes this db seed implementation to generate zero valid records when the fix for the error is added. It's a bit verbose to explain here so I will bring it up today on OH and we can document it that way.

**Comment by elipe17** [comment]
```
2024-09-09 18:12:27,335 ERROR case_consistency_validator.py::validate:L159 :  Uncaught exception during category four validation.
Traceback (most recent call last):
  File "/tdpapp/tdpservice/parsers/case_consistency_validator.py", line 156, in validate
    num_errors = self.__validate()
  File "/tdpapp/tdpservice/parsers/case_consistency_validator.py", line 176, in __validate
    num_errors += self.__validate_section2(num_errors)
  File "/tdpapp/tdpservice/parsers/case_consistency_validator.py", line 192, in __validate_section2
    num_errors += self.__validate_t5_atd_and_ssi()
  File "/tdpapp/tdpservice/parsers/case_consistency_validator.py", line 484, in __validate_t5_atd_and_ssi
    dob_date = datetime.strptime(dob, '%Y%m%d')
  File "/usr/local/lib/python3.10/_strptime.py", line 568, in _strptime_datetime
    tt, fraction, gmtoff_fraction = _strptime(data_string, format)
  File "/usr/local/lib/python3.10/_strptime.py", line 352, in _strptime
    raise ValueError("unconverted data remains: %s" %
ValueError: unconverted data remains: 08
```

**Comment by andrew-jameson** [comment]
> ```
> 2024-09-09 18:12:27,335 ERROR case_consistency_validator.py::validate:L159 :  Uncaught exception during category four validation.
> Traceback (most recent call last):
>   File "/tdpapp/tdpservice/parsers/case_consistency_validator.py", line 156, in validate
>     num_errors = self.__validate()
>   File "/tdpapp/tdpservice/parsers/case_consistency_validator.py", line 176, in __validate
>     num_errors += self.__validate_section2(num_errors)
>   File "/tdpapp/tdpservice/parsers/case_consistency_validator.py", line 192, in __validate_section2
>     num_errors += self.__validate_t5_atd_and_ssi()
>   File "/tdpapp/tdpservice/parsers/case_consistency_validator.py", line 484, in __validate_t5_atd_and_ssi
>     dob_date = datetime.strptime(dob, '%Y%m%d')
>   File "/usr/local/lib/python3.10/_strptime.py", line 568, in _strptime_datetime
>     tt, fraction, gmtoff_fraction = _strptime(data_string, format)
>   File "/usr/local/lib/python3.10/_strptime.py", line 352, in _strptime
>     raise ValueError("unconverted data remains: %s" %
> ValueError: unconverted data remains: 08
> ```

@elipe17 Are you fine resolving this in follow on #3177 ?

**Review by raftmsohani — APPROVED** [review]

**Comment by elipe17** [comment]
> > ```
> > 2024-09-09 18:12:27,335 ERROR case_consistency_validator.py::validate:L159 :  Uncaught exception during category four validation.
> > Traceback (most recent call last):
> >   File "/tdpapp/tdpservice/parsers/case_consistency_validator.py", line 156, in validate
> >     num_errors = self.__validate()
> >   File "/tdpapp/tdpservice/parsers/case_consistency_validator.py", line 176, in __validate
> >     num_errors += self.__validate_section2(num_errors)
> >   File "/tdpapp/tdpservice/parsers/case_consistency_validator.py", line 192, in __validate_section2
> >     num_errors += self.__validate_t5_atd_and_ssi()
> >   File "/tdpapp/tdpservice/parsers/case_consistency_validator.py", line 484, in __validate_t5_atd_and_ssi
> >     dob_date = datetime.strptime(dob, '%Y%m%d')
> >   File "/usr/local/lib/python3.10/_strptime.py", line 568, in _strptime_datetime
> >     tt, fraction, gmtoff_fraction = _strptime(data_string, format)
> >   File "/usr/local/lib/python3.10/_strptime.py", line 352, in _strptime
> >     raise ValueError("unconverted data remains: %s" %
> > ValueError: unconverted data remains: 08
> > ```
> 
> @elipe17 Are you fine resolving this in follow on #3177 ?

@andrew-jameson Ya I think it makes sense to discuss/fix the cat4 bug in a separate ticket and leave the case consistent DB seed work as the ticket you linked.

**Review by elipe17 — APPROVED** [review]

**Pull request merged**